LUCENE-8875: Introduce Optimized Collector For Large Number Of Hits

atris opened the pull request:
Existing top score docs collectors prepopulate the priority queue
with sentinel objects, which can be expensive for large number
of hits requested. This commit introduces a new Collector which
optimizes the said case

atris commented:
I ran some tests with N requested as 100K but the matching docs being only around 10K. This gives a sizeable improvement in the overall size of the priority queue.

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class LargeNumHitsTopDocsCollector extends TopScoreDocCollector {
+
+  LargeNumHitsTopDocsCollector(int numHits, int totalHitsThreshold) {
```
users won't be able to use it if it doesn't have a public constructor
also I don't think we need to pass a totalHitsThreshold: the fact that lots of hits need to be fetched would disable optimizations to the retrieval of top hits anyway

atris replied:
+1, will fix

atris commented:
> Not prepopulating the hit queue is only one part of the problem, we would also need to not allocate `numHits` slots in the priority queue right away?
>
That would require a different PriorityQueue implementation, since the default implementation always allocates numHits slots and sets maxSize to numHits. We would need an implementation that only allocates two slots initially, but has the ability to "resize" as needed. WDYT?

> I'd rather like that we don't do any change to TopDocsCollector and not try to extend it, in my opinion those collectors are solving a very different problem and would need to evolve independently.

My main objective to use TopScoreDocsCollector was to reuse code as much as I could -- but I agree with your point. Will change the implementation and keep TopScoreDocsCollector pristine.

jpountz commented:
> That would require a different PriorityQueue implementation

Actually I don't think we need a growable priority queue. For such large number of hits it'd be probably more efficient to collect hits in an ArrayList first and only turn it into a PQ once there are `numHits` hits?

Pinging @tokee since this is a topic he already spent time thinking about. :)

atris commented:
> Actually I don't think we need a growable priority queue. For such large number of hits it'd be probably more efficient to collect hits in an ArrayList first and only turn it into a PQ once there are `numHits` hits?

Would that mean collecting all hits in the ArrayList, building a heap out of them iteratively once we exhaust documents and then calling top() numHits times?

tokee commented:
I don't know if I have much to add to https://sbdevel.wordpress.com/2015/10/05/speeding-up-core-search/ that Atri linked to on the JIRA: Yes, it's definitely possible to do tricks with large result sets.  Especially with the simple "just sort on score"-case where the really large wins in my book is less GC pressure by using a single `long[]` to hold the structure instead of a gazillion small objects.

Here's the but: I haven't pursued it further as we (Royal Danish Library) have little use for it. Being able to handle large result sets in a single shards does not help much with multi-shard setups, where the merging node is likely to blow up. An iterative approach, such as `cursorMark` or Solr's `export`, is less prone to surprises and scales indefinitely. That being said, I won't stand in the way of building a fine foot blowing gun - for some use cases it would be a great win.

atris commented:
@jpountz @tokee Thanks for your comments.

I am planning to maintain an ArrayList of ScoreDocs and collect numHits, then do a sort on score and return top N as Adrien suggested. This should optimize the performance bottleneck around PQ allocating slots and prepopulating sentinel values.

I am inclined to pursue the idea of not using ScoreDoc and representing score + docID as a separate issue primarily due to the nuances involved with handling shard indices. I do feel that should not be a major blocker anymore given the fact that TopDocs.merge can now tie break on docIDs and ignore shard indices completely during tie breaking.

Thoughts?

atris commented:
@jpountz I have pushed a new iteration which does as discussed i.e. builds a hits list and populates hits as long as the number of collected hits are lesser than number of hits requested. Once that threshold is reached, a priority queue is built and minimum competitive score is set and used to filter further hits.

Please let me know if it looks fine.

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,177 @@
+          pq = new HitQueue(numHits, false);
+
+          for (int i = 0; i < hits.size(); i++) {
+            pq.add(hits.get(i));
```
use `for (ScoreDoc scoreDoc : hits)`  instead?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,177 @@
+      scorer.setMinCompetitiveScore(Math.nextUp(pqTop.score));
+    }
+    totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+  }
```
Maybe remove this logic, this is unlikely to help when fetching lots of hits.

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,177 @@
+
+        return o1.doc > o2.doc ? 1 : -1;
+      }
+    });
```
`Comparator.comparing(scoreDoc -> scoreDoc.score).reversed().thenComparing(scoreDoc -> scoreDoc.doc)`?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,177 @@
+      @Override
+      public int compare(ScoreDoc o1, ScoreDoc o2) {
+        if (o1.score < o2.score) {
+          return -1;
```
This seems to be putting lower scores first, while we'd like the higher scores first?

atris replied:
Removed the comparator

atris commented:
@jpountz Thanks for the comments, update the PR.

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,133 @@
+
+      assertEquals(firstScoreDoc.doc, secondScoreDoc.doc);
+      assertEquals(firstScoreDoc.score, secondScoreDoc.score, 0);
+    }
```
You can replace this look with CheckHits#checkEquals.

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  private void runNumHits(int numHits) throws IOException {
+    Query q = new MatchAllDocsQuery();
```
Maybe use a query that doesn't produce constant scores to make sure that the collector orders by score? For instance you could add a term to some documents and then build a disjunction here across a query on this term and a MatchAllDocsQuery?

atris replied:
+1, updated

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,133 @@
+
+    expectThrows(IllegalArgumentException.class, () -> {
+      largeCollector.topDocs(350_000);
+    });
```
can you check the error message?

jpountz commented:
@atris FYI avoiding force pushes would be helpful to reviewers as we could then look at what exactly changed compared to the previous PR.

atris commented:
@jpountz  I have updated the PR per your comments. ant precommit passes.

Apologies, this iteration also got force pushed. I have a local daemon which auto squashes and force pushes to my fork each time I create a new commit on a branch (this was meant to help the committer merge the PR without needing to squash). I will disable that in the next runs.

atris commented:
FYI ant precommit passes -- ran it with latest iteration

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,157 @@
+  ScoreDoc pqTop;
+  int totalHits;
+  /** Whether {@link #totalHits} is exact or a lower bound. */
+  protected TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
```
No need for `protected` since since class is not designed to be extended? Maybe we should make it final by the way?

atris replied:
That was something I was not sure about -- should we allow subclasses for this one? I agree with your line of thinking, will make it final

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,157 @@
+  private List<ScoreDoc> hits = new ArrayList<>();
+  HitQueue pq;
+  ScoreDoc pqTop;
+  int totalHits;
```
make these members private?

atris replied:
I am letting pq and pqTop be package private to allow tests to verify correctness

jpountz replied:
Ah I see it now. Can you add "// pkg-private for testing" on top? This is usually what we do when we need to do this.

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,163 @@
+    return results == null ? EMPTY_TOPDOCS : new TopDocs(new TotalHits(totalHits, totalHitsRelation), results);
+  }
+
+  public TopDocs topDocs() {
```
Add javadocs?

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,141 @@
+        doc.add(newStringField("field", "5", Field.Store.NO));
+        writer.addDocument(doc);
+      }
+      else {
```
we usually put the else statement on the same line as the closing bracket of the if statement

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,141 @@
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(250_000, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
```
can you use testQuery instead to make sure that the list is sorted before creating top docs?

jpountz commented on lucene/sandbox/src/test/org/apache/lucene/search/TestLargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,141 @@
+    TopScoreDocCollector regularCollector = TopScoreDocCollector.create(100_000, null, Integer.MAX_VALUE);
+
+    searcher.search(q, largeCollector);
+    searcher.search(q, regularCollector);
```
use testQuery here as well?

atris commented:
@jpountz Thanks, I have fixed the comments. The reason I did not use the BooleanQuery for the PQ build and not build tests was because those tests only exercise whether the priority queue was built or not which is dependent on the number of hits collected vs number of hits requested. However, per your suggestion, I have fixed the same.

atris commented:
Added a dedicated test for checking order of hits when TopDocs are returned from just the hits list and no PQ is built

jpountz commented:
Have you run tests? `TestLargeNumHitsTopDocsCollector.testLargeNumAndSparseHits` fails all the time for me.

jpountz commented:
Also tests run in several seconds right now, which is too slow. Can we make them run faster?

mikemccand commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,159 @@
+  private final int numHits;
+  private List<ScoreDoc> hits = new ArrayList<>();
+  /** Whether {@link #totalHits} is exact or a lower bound. */
+  private TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
```
Why do we need a member for this?  It's never changed?  Can we just inline `TotalHits.Relation.EQUAL_TO` below?

atris replied:
Sure, I kept it here to increase readability -- will inline

mikemccand commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,159 @@
+  int totalHits;
+
+  public LargeNumHitsTopDocsCollector(int numHits) {
+    this.numHits = numHits;
```
Can we rename `numHits` to `requestedHitCount` or so?  Otherwise `numHits` and `totalHits` are confusable?

mikemccand commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,159 @@
+          // Convert the list to a priority queue
+
+          // We should get here only when priority queue
+          // has been built
```
has `not yet` been built you mean?

atris replied:
Good catch, thanks :)

atris commented:
@jpountz @mikemccand Thanks for your comments, fixed the same.

I am surprised that the test is failing : it seems to be consistently passing for me. I ran a couple of times.

I have also reduced the dataset for tests: there is no real functional difference between 100 and 250k documents for testing, since these are not performance tests.

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java:
```diff
@@ -0,0 +1,157 @@
+        assert score >= 0; // NOTE: false for NaN
+
+        if (totalHits < requestedHitCount) {
+          hits.add(new ScoreDoc(doc, score));
```
OK I dug the test failure, it was due to the fact that you are passing `doc` here instead of `docBase + doc`. I'm curious why you couldn't find it, it happened pretty often for me.

jpountz commented:
> Too bad we cannot share any (of this complex) code with existing collectors

The initial PR tried to share some logic, but I suggested to avoid doing this as I was expecting this collector to work fairly differently given that it operates on different assumptions. Looking at the current code, it doesn't seem to be like there is a lot to be shared?

> I am surprised that the test is failing : it seems to be consistently passing for me. I ran a couple of times.

I identified the bug, it was due to collecting `doc` instead of `docBase+doc` when collecting into an ArrayList. Maybe there is something to be refactored in the test so that it would have surfaced this issue more consistently?

atris commented:
@jpountz Thanks for looking through!

I refactored the tests to remove the dependency on random() to populate the string Field.  That triggered the test for me in the first run. Maybe something to do with the seed?

atris commented:
I just discovered ant beast that seems like the right tool for ensuring that such random tests get reproduced and are not dependent on lucky/unlucky seeds. Further PRs will have ant beast run on the new random tests before I raise a PR

wuyunfeng commented:
@atris  This PR is just resolve the cost for  prepopulating the priority queue with sentinel objects?   So amazing for the cost for  prepopulating.

atris commented:
Not just the cost of prepopulating the PQ, but also building it in first place. If total number of hits < number of requested hits and number of requested hits is  a large value, then it makes sense to skip the PQ

wuyunfeng commented:
@atris  Can you give some result comparison or other evidence?

atris commented:
> @atris Can you give some result comparison or other evidence?

Please see the blog post listed above.

In addition, the targeted use case for this collector makes the benefits obvious: I ran a TermQuery on a 5 TB dataset manufactured from Elasticsearch's nyc_taxis dataset. A string field was modified to have a specific value only for 50,000 documents, and number of hits requested was 25,00,000 by each of 30 concurrent requests (the same test mentioned in Toke's blogpost). The space savings were in order of multiple GBs, as expected.

atris commented:
@jpountz JFYI I ran ant beast in 3 batches of 10 times each and it ran clean:

WARNING: All illegal access operations will be denied in a future release
  [beaster] Beast round 1 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/1
  [beaster] Beast round 2 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/2
  [beaster] Beast round 3 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/3
  [beaster] Beast round 4 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/4
  [beaster] Beast round 5 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/5
  [beaster] Beast round 6 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/6
  [beaster] Beast round 7 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/7
  [beaster] Beast round 8 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/8
  [beaster] Beast round 9 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/9
  [beaster] Beast round 10 results: /Users/atris/merge_refactor/lucene-solr/lucene/build/sandbox/test/10
  [beaster] Beasting finished Successfully.

-check-totals:

beast:

BUILD SUCCESSFUL

wuyunfeng commented:
@atris Hi，atris.  which Lucene release version would use this PR？I hope i can use this.
